src/python/pants/engine/target.py

```diff
@@ -124,20 +101,37 @@ def compute_value(cls, raw_value: Optional[int], *, address: Address) -> Optiona
                 return value_or_default
     """
 
-    value: ImmutableValue
+    # This exists on every Field subclass.
+    value: _DefaultBase
+    # Subclasses must define these.
+    alias: ClassVar[str]
+    default: ClassVar[_DefaultBase]
+    # Subclasses may define these.
+    # TODO: enforce that `raw_value` and `value` are *not* Optional if `required = True`, possibly
+    # via mixin or decorator?
+    required: ClassVar[bool] = False
+    deprecated_removal_version: ClassVar[Optional[str]] = None
+    deprecated_removal_hint: ClassVar[Optional[str]] = None
 
+    # This method is marked @final so that we have a common deprecation mechanism, and to ensure
+    # that all subclasses have this exact type signature for their constructor.
```

Comment:
This is a bit stale now. I think it's fine to drop the comment.

It's `@final` because we don't expect subclasses to be mucking around with the constructor; they should change `compute_value()` instead. But I'm not sure we need to explain that, as `@final` already expresses we don't want subclasses changing things.